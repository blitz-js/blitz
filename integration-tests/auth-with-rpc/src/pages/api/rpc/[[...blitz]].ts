import {rpcHandler} from "@blitzjs/rpc"
import {api} from "../../../blitz-server"

export default api(rpcHandler({onError: console.log}))
