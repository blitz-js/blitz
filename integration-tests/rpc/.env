VITE_BLITZ_TEST_ENVIRONMENT=true