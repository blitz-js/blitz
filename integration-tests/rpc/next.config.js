const {withBlitz} = require("@blitzjs/next")
module.exports = withBlitz({
  target: "experimental-serverless-trace",
})
