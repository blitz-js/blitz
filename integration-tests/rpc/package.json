{
  "name": "test-rpc",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "test": "vitest --config ./vitest.config.ts run",
    "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf .next"
  },
  "dependencies": {
    "@blitzjs/config": "workspace:2.0.0-beta.21",
    "@blitzjs/next": "workspace:2.0.0-beta.21",
    "@blitzjs/rpc": "workspace:2.0.0-beta.21",
    "blitz": "workspace:2.0.0-beta.21",
    "next": "12.2.5",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "@types/express": "4.17.13",
    "@types/fs-extra": "9.0.13",
    "@types/node-fetch": "2.6.1",
    "@types/react": "18.0.25",
    "b64-lite": "1.4.0",
    "eslint": "8.27.0",
    "fs-extra": "10.0.1",
    "typescript": "^4.8.4"
  }
}
