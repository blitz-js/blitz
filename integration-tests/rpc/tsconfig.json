{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "incremental": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "plugins": [
      {
        "name": "next"
      }
    ],
    "strictNullChecks": true
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", "types", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}
