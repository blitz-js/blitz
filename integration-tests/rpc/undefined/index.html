<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <meta name="next-head-count" content="2" />
    <noscript data-n-css=""></noscript>
    <script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script>
    <script src="/_next/static/chunks/webpack-883db370285692a4.js" defer=""></script>
    <script src="/_next/static/chunks/framework-9b5cc288ac193d39.js" defer=""></script>
    <script src="/_next/static/chunks/main-04c3f0a1123ec3c0.js" defer=""></script>
    <script src="/_next/static/chunks/pages/_app-951c3594bcece24b.js" defer=""></script>
    <script src="/_next/static/chunks/pages/index-81cb1927ec0e20ed.js" defer=""></script>
    <script src="/_next/static/pCckmZw1vnHszTyFb4kqd/_buildManifest.js" defer=""></script>
    <script src="/_next/static/pCckmZw1vnHszTyFb4kqd/_ssgManifest.js" defer=""></script>
  </head>
  <body>
    <div id="__next"><div id="page-container">Hello World</div></div>
    <script id="__NEXT_DATA__" type="application/json">
      {
        "props": {"pageProps": {}},
        "page": "/",
        "query": {},
        "buildId": "pCckmZw1vnHszTyFb4kqd",
        "runtimeConfig": {},
        "nextExport": true,
        "autoExport": true,
        "isFallback": false,
        "scriptLoader": []
      }
    </script>
  </body>
</html>
