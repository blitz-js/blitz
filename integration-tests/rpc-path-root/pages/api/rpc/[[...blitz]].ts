import {rpcHandler} from "@blitzjs/rpc"

export default rpcHandler({onError: console.log})
