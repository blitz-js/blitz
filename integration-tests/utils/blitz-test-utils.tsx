import React from "react"
import {vi} from "vitest"
import {QueryClient} from "@tanstack/react-query"
import {BlitzRpcPlugin, QueryClientProvider} from "@blitzjs/rpc"
import {NextRouter} from "next/router"
import {RouterContext} from "next/dist/shared/lib/router-context"
import {render as defaultRender} from "@testing-library/react"

export const mockRouter: NextRouter = {
  basePath: "",
  pathname: "/",
  route: "/",
  asPath: "/",
  query: {},
  isReady: true,
  isLocaleDomain: false,
  isPreview: false,
  push: vi.fn(),
  replace: vi.fn(),
  reload: vi.fn(),
  back: vi.fn(),
  prefetch: vi.fn(),
  beforePopState: vi.fn(),
  events: {
    on: vi.fn(),
    off: vi.fn(),
    emit: vi.fn(),
  },
  isFallback: false,
}

type DefaultParams = Parameters<typeof defaultRender>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & {
  router?: Partial<NextRouter>
}
export type BlitzProviderProps = {
  client?: QueryClient
  contextSharing?: boolean
}

const BlitzProvider = ({
  client,
  contextSharing = false,
  children,
}: BlitzProviderProps & {children: JSX.Element}) => {
  if (globalThis.queryClient) {
    return (
      <QueryClientProvider
        client={client || globalThis.queryClient}
        contextSharing={contextSharing}
      >
        {children}
      </QueryClientProvider>
    )
  }

  return children
}
const compose =
  (...rest) =>
  (x: React.ComponentType<any>) =>
    rest.reduceRight((y, f) => f(y), x)

const BlitzWrapper = ({plugins, children}) => {
  const providers = plugins.reduce((acc, plugin) => {
    return plugin.withProvider ? acc.concat(plugin.withProvider) : acc
  }, [])
  const withPlugins = compose(...providers)
  const component = React.useMemo(() => withPlugins(children), [children])

  return (
    <BlitzProvider>
      <RouterContext.Provider value={{...mockRouter}}>{component}</RouterContext.Provider>
    </BlitzProvider>
  )
}

export function render(ui: RenderUI, {wrapper, router, ...options}: RenderOptions = {}) {
  if (!wrapper) {
    wrapper = ({children}) => {
      return <BlitzWrapper plugins={[BlitzRpcPlugin({})]}>{children}</BlitzWrapper>
    }
  }

  return defaultRender(ui, {wrapper, ...options})
}

// This enhance fn does what buildRpcFunction does during build time
export function buildQueryRpc(fn: any) {
  const newFn = (...args: any) => {
    const [data, ...rest] = args
    return fn(data, ...rest)
  }
  newFn._isRpcClient = true
  newFn._resolverType = "query"
  newFn._routePath = "/api/test/url/" + Math.random()
  return newFn
}

// This enhance fn does what buildRpcFunction does during build time
export function buildMutationRpc(fn: any) {
  const newFn = (...args: any) => fn(...args)
  newFn._isRpcClient = true
  newFn._resolverType = "mutation"
  newFn._routePath = "/api/test/url"
  return newFn
}
