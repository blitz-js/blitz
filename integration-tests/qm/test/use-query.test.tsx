import {describe, it, expect, beforeAll, vi} from "vitest"
import {act, screen, waitForElementToBeRemoved, waitFor} from "@testing-library/react"
import {useQuery, useInfiniteQuery, BlitzRpcPlugin, QueryClientProvider} from "@blitzjs/rpc"
import React from "react"
import delay from "delay"
import {buildMutationRpc, buildQueryRpc, mockRouter, render} from "../../utils/blitz-test-utils"
import {RouterContext} from "next/dist/shared/lib/router-context"

beforeAll(() => {
  globalThis.__BLITZ_SESSION_COOKIE_PREFIX = "qm-test-cookie-prefix"
  globalThis.IS_REACT_ACT_ENVIRONMENT = true
})

describe("useQuery", () => {
  const setupHook = (
    ID: string,
    params: any,
    queryFn: (...args: any) => any,
    options: Parameters<typeof useQuery>[2] = {} as any,
  ): [{data?: any; setQueryData?: any}, Function] => {
    let res = {}
    const qc = BlitzRpcPlugin({})

    function TestHarness() {
      const [data, {setQueryData}] = useQuery(queryFn, params, {
        suspense: true,
        ...options,
      } as any)

      Object.assign(res, {data, setQueryData})
      return (
        <div id={`harness-${ID}`}>
          <span>{data ? `Ready${ID}` : "No data"}</span>
          <span>{data}</span>
        </div>
      )
    }

    const ui = () => (
      <React.Suspense fallback="Loading...">
        <TestHarness />
      </React.Suspense>
    )

    const {rerender} = render(ui(), {
      wrapper: ({children}) => (
        <QueryClientProvider client={globalThis.queryClient}>
          <RouterContext.Provider value={mockRouter}>{children}</RouterContext.Provider>
        </QueryClientProvider>
      ),
    })
    return [res, () => rerender(ui())]
  }

  describe('a "query" that converts the string parameter to uppercase', () => {
    const upcase = async (args: string) => {
      await delay(500)
      return args.toUpperCase()
    }

    it("should work with Blitz queries", async () => {
      const [res] = setupHook("2", "test", buildQueryRpc(upcase))
      await waitForElementToBeRemoved(() => screen.getByText("Loading..."))
      await act(async () => {
        await screen.findByText("Ready2")
        expect(res.data).toBe("TEST")
      })
    })

    it("should be able to change the data with setQueryData", async () => {
      const [res] = setupHook("3", "fooBar", buildQueryRpc(upcase))
      await waitForElementToBeRemoved(() => screen.getByText("Loading..."))
      await act(async () => {
        await screen.findByText("Ready3")
        expect(res.data).toBe("FOOBAR")
        res.setQueryData((p: string) => p.substr(3, 3), {refetch: false})
        await delay(100)
      })

      expect(res.data).toBe("BAR")
    })

    it("shouldn't work with regular functions", () => {
      console.error = vi.fn()
      expect(() => setupHook("4", "test", upcase)).toThrowErrorMatchingSnapshot()
    })

    it("shouldn't work with mutation function", () => {
      console.error = vi.fn()
      expect(() => setupHook("5", "test", buildMutationRpc(upcase))).toThrowErrorMatchingSnapshot()
    })

    it("suspense disabled if enabled is false", async () => {
      setupHook("6", "test", buildQueryRpc(upcase), {enabled: false})
      await screen.findByText("No data")
    })

    it("suspense disabled if enabled is undefined", async () => {
      setupHook("7", "test", buildQueryRpc(upcase), {enabled: undefined})
      await screen.findByText("No data")
    })

    // it("suspense disabled if enabled is false and suspense set", async () => {
    //   setupHook("8", "test", buildQueryRpc(upcase), {
    //     enabled: false,
    //     suspense: true,
    //   })
    //   await screen.findByText("No data")
    // })

    it("works with options other than enabled & suspense without type error", () => {
      const Demo = () => {
        useQuery(buildQueryRpc(upcase), undefined, {refetchInterval: 10000})
        return <div></div>
      }
      const ui = () => <Demo />

      const {rerender} = render(ui(), {
        wrapper: ({children}) => (
          <QueryClientProvider client={globalThis.queryClient}>
            <RouterContext.Provider value={mockRouter}>{children}</RouterContext.Provider>
          </QueryClientProvider>
        ),
      })
    })
  })
})

describe("useInfiniteQuery", () => {
  const setupHook = (
    ID: string,
    params: (arg?: any) => any,
    queryFn: (...args: any) => any,
  ): [{data?: any; setQueryData?: any}, Function] => {
    let res = {}
    const qc = BlitzRpcPlugin({})

    function TestHarness() {
      // TODO - fix typing
      //@ts-ignore
      const [groupedData] = useInfiniteQuery(queryFn, params, {
        suspense: true,
        getNextPageParam: () => {},
      })
      Object.assign(res, {groupedData})
      return (
        <div id="harness">
          <span>{groupedData ? `Ready${ID}` : "No data"}</span>
          <div>
            {groupedData.map((data: any, i) => (
              <div key={i}>{data}</div>
            ))}
          </div>
        </div>
      )
    }

    const ui = () => (
      <React.Suspense fallback={`Loading${ID}...`}>
        <TestHarness />
      </React.Suspense>
    )

    const {rerender} = render(ui(), {
      wrapper: ({children}) => (
        <QueryClientProvider client={globalThis.queryClient}>
          <RouterContext.Provider value={mockRouter}>{children}</RouterContext.Provider>
        </QueryClientProvider>
      ),
    })
    return [res, () => rerender(ui())]
  }

  const getItems = ({id}: {id: number}) => {
    if (id === 1) {
      return "item1"
    } else if (id === 2) {
      return "item2"
    } else {
      throw new Error("No item for this id")
    }
  }
  it("should show loading", async () => {
    setupHook("1", () => ({id: 1}), buildQueryRpc(getItems))
    await waitForElementToBeRemoved(() => screen.getByText("Loading1..."))
    await act(async () => {
      await screen.findByText("item1")
    })
  })
})
