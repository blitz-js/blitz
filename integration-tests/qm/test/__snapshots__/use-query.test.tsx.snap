// Vitest Snapshot v1

exports[`useQuery > a "query" that converts the string parameter to uppercase > shouldn't work with mutation function 1`] = `"Cannot read properties of null (reading 'isReady')"`;

exports[`useQuery > a "query" that converts the string parameter to uppercase > shouldn't work with regular functions 1`] = `"Cannot read properties of null (reading 'isReady')"`;
